### PR TITLE
feat(CalibrationStatusClassification): improve preprocessing & rework config model

### PR DESCRIPTION
**Pull request by amadeuszsz**
## Summary

Update Calibration Status Classification model's config, improve preprocessing and tiny refactor.

## Change point
* Speedup preprocessing pipeline. From OpenCV per-pixel loop processing to numpy based array dilation with improvement from about 3500 ms to 30 ms.
* Dropping out of range LiDAR points. Before, we set all the points beyond the given range as this range, which generated fake data.
* Function for generating binary data after preprocessing. When path set, it saves raw data. It can be further utilized for unit tests in order to validate preprocessing output in target environment (ROS 2 node).
* Model's config rework.

## Note

For easier review purpose, model's config name is unchanged. Before merge we need to change config name to `resnet18_5ch_1xb16-50e_j6gen2.py`.

## Test performed

- Test log
```
2025/09/01 05:09:26 - mmengine - INFO - Epoch(test) [2714/2714]    accuracy/top1: 99.8180  single-label/precision: 99.8177  single-label/recall: 99.8186  single-label/f1-score: 99.8180  data_time: 5.8287  time: 6.1218
```

- Deployment log
```
INFO:mmdeploy:Exporting model to ONNX format...
INFO:mmdeploy:Input shape: torch.Size([1, 5, 1860, 2880])
INFO:mmdeploy:Output path: /workspace/work_dirs/end2end.onnx
INFO:mmdeploy:ONNX opset version: 16
INFO:mmdeploy:ONNX export completed: /workspace/work_dirs/end2end.onnx
INFO:mmdeploy:Simplifying ONNX model...
INFO:mmdeploy:ONNX model simplified successfully. Saved to /workspace/work_dirs/end2end.onnx
INFO:mmdeploy:Converting ONNX to TensorRT...
INFO:mmdeploy:Setting input shapes - min: [1, 5, 1080, 1920], opt: [1, 5, 1860, 2880], max: [1, 5, 2160, 3840]
INFO:mmdeploy:Successfully parsed the ONNX file
INFO:mmdeploy:Building TensorRT engine...
INFO:mmdeploy:TensorRT engine saved to /workspace/work_dirs/end2end.engine
INFO:mmdeploy:Engine max workspace size: 1.00 GB
INFO:mmdeploy:TensorRT conversion successful: /workspace/work_dirs/end2end.engine
INFO:mmdeploy:Running verification for miscalibrated and calibrated samples with an output array [SCORE_MISCALIBRATED, SCORE_CALIBRATED]...
INFO:mmdeploy:==================================================
INFO:mmdeploy:Verifying miscalibrated sample...
INFO:mmdeploy:--------------------------------------------------
INFO:mmdeploy:Verifying PyTorch model...
INFO:mmdeploy:PyTorch inference latency: 1010.47 ms
INFO:mmdeploy:Output verification:
INFO:mmdeploy:  Output: [[ 1.4509633 -1.4554818]]
INFO:mmdeploy:PyTorch output for miscalibrated: [SCORE_MISCALIBRATED, SCORE_CALIBRATED] = [[ 1.4509633 -1.4554818]]
INFO:mmdeploy:Negative calibration score detected for miscalibrated sample: -2.906445
INFO:mmdeploy:--------------------------------------------------
INFO:mmdeploy:Verifying ONNX model...
INFO:mmdeploy:ONNX inference latency: 1431.94 ms
INFO:mmdeploy:Output verification:
INFO:mmdeploy:  ONNX output: [[ 1.4509625 -1.455481 ]]
INFO:mmdeploy:  Max difference with PyTorch: 0.000001
INFO:mmdeploy:  Mean difference with PyTorch: 0.000001
INFO:mmdeploy:ONNX model verification passed!
INFO:mmdeploy:--------------------------------------------------
INFO:mmdeploy:Verifying TensorRT model...
INFO:mmdeploy:TensorRT inference latency: 6.41 ms
INFO:mmdeploy:Output verification:
INFO:mmdeploy:  TensorRT output: [[ 1.450984  -1.4555026]]
INFO:mmdeploy:  Max difference with PyTorch: 0.000021
INFO:mmdeploy:  Mean difference with PyTorch: 0.000021
INFO:mmdeploy:TensorRT model verification passed!
INFO:mmdeploy:==================================================
INFO:mmdeploy:Verifying calibrated sample...
INFO:mmdeploy:--------------------------------------------------
INFO:mmdeploy:Verifying PyTorch model...
INFO:mmdeploy:PyTorch inference latency: 1024.37 ms
INFO:mmdeploy:Output verification:
INFO:mmdeploy:  Output: [[-0.95841575  0.950384  ]]
INFO:mmdeploy:PyTorch output for calibrated: [SCORE_MISCALIBRATED, SCORE_CALIBRATED] = [[-0.95841575  0.950384  ]]
INFO:mmdeploy:Positive calibration score detected for calibrated sample: 1.908800
INFO:mmdeploy:--------------------------------------------------
INFO:mmdeploy:Verifying ONNX model...
INFO:mmdeploy:ONNX inference latency: 577.60 ms
INFO:mmdeploy:Output verification:
INFO:mmdeploy:  ONNX output: [[-0.95841527  0.950384  ]]
INFO:mmdeploy:  Max difference with PyTorch: 0.000000
INFO:mmdeploy:  Mean difference with PyTorch: 0.000000
INFO:mmdeploy:ONNX model verification passed!
INFO:mmdeploy:--------------------------------------------------
INFO:mmdeploy:Verifying TensorRT model...
INFO:mmdeploy:TensorRT inference latency: 6.39 ms
INFO:mmdeploy:Output verification:
INFO:mmdeploy:  TensorRT output: [[-0.9584862   0.95045495]]
INFO:mmdeploy:  Max difference with PyTorch: 0.000071
INFO:mmdeploy:  Mean difference with PyTorch: 0.000071
INFO:mmdeploy:TensorRT model verification passed!
INFO:mmdeploy:==================================================
INFO:mmdeploy:Deployment completed successfully!
```

## Artifacts

- Checkpoint - [TIER IV INTERNAL LINK](https://drive.google.com/file/d/1Iyhf6PELT7ht-YvHC4ESi_zPSaRgeSbA/view?usp=drive_link)
- ONNX - [TIER IV INTERNAL LINK](https://drive.google.com/file/d/1t0f2W1VHxMraYJO602mpaOT2_MojjKfw/view?usp=drive_link)
- Train log - [TIER IV INTERNAL LINK](https://drive.google.com/file/d/1JlcAugS5i6JW9BF7cq-M6LuRlYsiZV8q/view?usp=drive_link)
- Test log - [TIER IV INTERNAL LINK](https://drive.google.com/file/d/1aqzcn9FlD6g4O7Bbgot4ewmXxPERd7Hp/view?usp=drive_link)

### CODE REVIEW

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py [hunk 1]
```diff
@@ -584,7 +671,7 @@ def _generate_input_data(
         if augmentation_tf is not None:
             pointcloud_ics = self._apply_augmentation_to_points(pointcloud_ics, augmentation_tf)
 
-        return self._create_lidar_images(image, pointcloud_ics, pointcloud_ccs, intensities)
+        return self._create_lidar_images(image, pointcloud_ics, pointcloud_ccs, intensities).astype(np.float32) / 255.0
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py:673]: Addressed in https://github.com/tier4/AWML/pull/87/commits/bcade7569aaec0d304e9b8e66480b340d8cdb256

[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py:673]: @vividf I see the error, sorry for trouble. Fixed with https://github.com/tier4/AWML/pull/87/commits/968b3b629c20aa56e61cc7ab74a0782440fae71d:
```
root@42fffe43ef50:/workspace# python tools/calibration_classification/visualize_lidar_camera_projection.py projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py --info_pkl data/t4dataset/calibration_info/t4dataset_gen2_base_infos_test.pkl --data_root data/t4dataset --output_dir ./calibration_visualization --sample_idx 102
09/04 06:57:45 - mmengine - INFO - Initialized CalibrationClassificationTransform with mode: test
09/04 06:57:45 - mmengine - INFO - Transform config: {'crop_ratio': 0.6, 'max_distortion': 0.001, 'rgb_weight': 0.3}
09/04 06:57:46 - mmengine - INFO - Loaded 43413 samples from info.pkl
09/04 06:57:46 - mmengine - INFO - Processing sample 102 from 43413 total samples
09/04 06:57:47 - mmengine - INFO - Saved test projection visualization to ./calibration_visualization/projection_test_sample_102_00017_CAM_FRONT_gt_label_1.png
09/04 06:57:47 - mmengine - INFO - Successfully processed sample 102
09/04 06:57:47 - mmengine - INFO - Visualizations saved to directory: ./calibration_visualization
```

File: autoware_ml/calibration_classification/hooks/result_visualization_hook.py [hunk 2]
```diff
@@ -222,7 +222,6 @@ def _visualize_results(
                     f"[ResultVisualizationHook] fused_img data not found for {img_path}, skipping visualization."
                 )
                 return
-
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/calibration_classification/hooks/result_visualization_hook.py]: Agree https://github.com/tier4/AWML/pull/87/commits/4a14e0c71f4e33744a0a84b78710e7c79abfcbdb

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py [hunk 3]
```diff
@@ -204,8 +208,98 @@ def transform(self, results: Dict[str, Any], force_generate_miscalibration: bool
...
+
+        # Save data
+        suffix = "calibrated" if label == 1 else "miscalibrated"
+        directory = f"sample_{idx}"
+        save_dir = os.path.join(self.binary_save_dir, directory)
+        os.makedirs(save_dir, exist_ok=True)
+        cv2.imwrite(os.path.join(save_dir, "image.png"), image)
+        cv2.imwrite(os.path.join(save_dir, "image_undistorted.png"), image_undistorted)
+        save_gzipped_numpy(image, os.path.join(save_dir, "image.dat.gz"))
+        save_gzipped_numpy(image_undistorted, os.path.join(save_dir, "image_undistorted.dat.gz"))
```
Comment (KSeangTan) [inline autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py:290]: A lot of repetition, can we make inputs a pair and do it it a for-loop?

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py [hunk 4]
```diff
@@ -666,23 +752,61 @@ def _create_lidar_images(
...
-                continue
+            sorted_rows = final_rows[sort_indices]
+            sorted_cols = final_cols[sort_indices]
+            sorted_depths = final_depths[sort_indices]
+            sorted_intensities = final_intensities[sort_indices]
 
-            distance_color = int(np.clip(255 * point3d[2] / depth_scale, 0, 255))
-            cv2.circle(depth_image, tuple(point2d.astype(int)), radius, distance_color, -1)
-            cv2.circle(intensity_image, tuple(point2d.astype(int)), radius, int(intensity), -1)
+            depth_image[sorted_rows, sorted_cols] = sorted_depths.astype(np.float32)
```
Comment (KSeangTan) [inline autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py:804]: is it necessary? `depth_image` and `intensity_image` are defined as `np.float32` at the beginning right?

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py [hunk 5]
```diff
@@ -666,23 +752,61 @@ def _create_lidar_images(
...
+            intensity_image[sorted_rows, sorted_cols] = sorted_intensities.astype(np.float32)
 
         depth_image = np.expand_dims(depth_image, axis=2)
         intensity_image = np.expand_dims(intensity_image, axis=2)
-        return np.concatenate([image, depth_image, intensity_image], axis=2)
+        return np.concatenate([image.astype(np.float32), depth_image, intensity_image], axis=2)
```
Comment (KSeangTan) [inline autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py:809]: It looks cleaner if we set `dtype` in `np.concatenate`

File: projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py [hunk 6]
```diff
@@ -30,30 +32,75 @@
...
         style="pytorch",
+        norm_cfg=dict(type="BN", requires_grad=True),
     ),
     neck=dict(type="GlobalAveragePooling"),
     head=dict(
         type="LinearClsHead",
         num_classes=2,
         in_channels=512,
-        loss=dict(type="CrossEntropyLoss", loss_weight=1.0),
+        loss=dict(
```
Comment (KSeangTan) [inline projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py:42]: Comment about why we use `LabelSmoothLoss` instead of `CrossEntropyLoss`?

File: projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py [hunk 7]
```diff
@@ -30,30 +32,75 @@
...
+        lr=0.001,
+        weight_decay=0.01,
+        betas=(0.9, 0.999),
+        eps=1e-8,
+    ),
+    clip_grad=dict(max_norm=2.0, norm_type=2),
+)
 
-train_cfg = dict(by_epoch=True, max_epochs=max_epochs, val_interval=1)
+param_scheduler = [
```
Comment (KSeangTan) [inline projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py:65]: Comment why we change to this scheduler?

File: projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py [hunk 8]
```diff
@@ -66,6 +113,8 @@
     num_workers=num_workers,
     persistent_workers=True,
     shuffle=True,
+    pin_memory=True,
+    drop_last=True,
```
Comment (KSeangTan) [inline projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py:117]: Any reason why we drop_last?

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py [hunk 9]
```diff
@@ -60,26 +61,32 @@ def __init__(
...
         """Initialize the CalibrationClassificationTransform.
         Args:
             transform_config (Dict[str, Any]): Configuration for transform parameters. Must contain:
                 - crop_ratio: Crop ratio for image augmentation (0.0, 1.0]
                 - max_distortion: Maximum distortion for affine transformation [0.0, 1.0]
-                - depth_scale: Scale factor for depth visualization (> 0.0)
-                - radius: Radius for point visualization (>= 0)
                 - rgb_weight: Weight for RGB overlay [0.0, 1.0]
             mode (str): Transform mode. Options: "train", "val", "test". Defaults to "train".
+            lidar_range (float): Range of LiDAR points which are considered valid. Defaults to 128.0.
```
Comment (KSeangTan) [inline autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py:81]: You meant euclidean distance in meters for lidar pointclouds?

**Review by KSeangTan — APPROVED** [review]
LGTM, please feel free to resolve comments and merge after you make changes

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py (hunk 4 above)
Comment (amadeuszsz) [reply autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py]: Actually adding list of pairs after proper code formatting will generate more lines of code.

Comment (amadeuszsz) [reply autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py]: Thanks! https://github.com/tier4/AWML/pull/87/commits/9a2e56194708bdb5b5a0c0d6582c430d1e9f5528

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py (hunk 5 above)
Comment (amadeuszsz) [reply autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py]: Thanks! https://github.com/tier4/AWML/pull/87/commits/994ffe71cf62ff4d0c36b8a8793b35284bb7cafb

File: autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py (hunk 9 above)
Comment (amadeuszsz) [reply autoware_ml/calibration_classification/datasets/transforms/calibration_classification_transform.py]: Good point! The name can be misleading. Changed to max_depth in https://github.com/tier4/AWML/pull/87/commits/d73a8ba9293d2727ced32721c5692d1f3a557a44

File: projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py (hunk 6 above)
Comment (amadeuszsz) [reply projects/CalibrationStatusClassification/configs/t4dataset/resnet18_5ch_1xb8-25e_j6gen2.py:42]: Considering the formula behind both loss functions and related task (calibration status classification) LabelSmoothLoss seems more intuitive way according to potential overconfidence. I did some experiments (about 50) related to all the changes in this PR and proposed configuration seems the best.
In future PR I would like to use optune, run experiments using this new config as a base and rework this config if necessary.
